dm/worker/server.go

```diff
@@ -480,9 +603,61 @@ OUTER:
 	return nil
 }
 
+func (s *Server) handleRelayConfig(ctx context.Context, relayCh chan ha.RelaySource, errCh chan error) error {
+OUTER:
+	for {
+		select {
+		case <-ctx.Done():
+			break OUTER
+		case relaySource, ok := <-relayCh:
+			if !ok {
+				break OUTER
+			}
+			log.L().Info("receive relay source", zap.String("relay source", relaySource.Source), zap.Bool("is deleted", relaySource.IsDeleted))
+			err := s.operateRelaySource(relaySource)
+			s.setSourceStatus(relaySource.Source, err, true)
+			if err != nil {
+				opErrCounter.WithLabelValues(s.cfg.Name, opErrTypeRelaySource).Inc()
+				log.L().Error("fail to operate relay source on worker",
+					zap.String("relay source", relaySource.Source),
+					zap.Bool("is deleted", relaySource.IsDeleted),
+					zap.Error(err))
+				if etcdutil.IsRetryableError(err) {
+					return err
+				}
+			}
+		case err, ok := <-errCh:
+			// currently no value is sent to errCh
+			if !ok {
+				break OUTER
+			}
+			// TODO: Deal with err
+			log.L().Error("WatchRelayConfig received an error", zap.Error(err))
+			if etcdutil.IsRetryableError(err) {
+				return err
+			}
+		}
+	}
+	log.L().Info("worker server is closed, handleRelayConfig will quit now")
+	return nil
+}
+
 func (s *Server) operateSourceBound(bound ha.SourceBound) error {
 	if bound.IsDeleted {
-		return s.stopWorker(bound.Source)
+		s.Lock()
+		w := s.getWorker(false)
```

Comment:
Why don't we directly `defer s.Unlock()` before this line?

Reply:
above comments should be addressed in https://github.com/pingcap/dm/pull/1511/commits/c212e43d8aa4b259672e752434343f68c5240e05